common/src/primitives/height.rs

```diff
@@ -27,11 +122,69 @@ impl BlockHeight {
         MAX
     }
 
-    pub fn inner(self) -> u64 {
-        self.0
+    pub fn checked_add(&self, rhs: HeightIntType) -> Option<Self> {
+        self.0.checked_add(rhs).map(Self::new)
+    }
+
+    pub fn next_height(&self) -> BlockHeight {
+        BlockHeight(self.0.checked_add(1).expect("Block height overflow"))
+    }
+}
+
+/////////////////////////////
+
+#[derive(Debug, Copy, Clone, PartialOrd, Ord, PartialEq, Eq, Encode, Decode)]
+pub struct BlockDistance(DistanceIntType);
+
+impl BlockDistance {
+    pub fn new(distance: DistanceIntType) -> Self {
+        Self(distance)
+    }
+}
+
+// Display should be defined for thiserr crate
+impl fmt::Display for BlockDistance {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl From<BlockDistance> for DistanceIntType {
+    fn from(block_height: BlockDistance) -> DistanceIntType {
+        block_height.0
     }
+}
+
+impl From<DistanceIntType> for BlockDistance {
+    fn from(w: DistanceIntType) -> BlockDistance {
+        BlockDistance(w)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn basic_arithmetic() {
+        let h1 = BlockHeight::new(5);
+        let d_4 = BlockDistance::new(4);
+        let d_5 = BlockDistance::new(5);
+        let d_6 = BlockDistance::new(6);
+        assert_eq!((h1 - d_4).unwrap(), BlockHeight::new(1));
+        assert_eq!((h1 - d_5).unwrap(), BlockHeight::new(0));
+        assert!((h1 - d_6).is_none());
+
+        assert_eq!((d_5 - d_4).unwrap(), BlockDistance::new(1));
+        assert_eq!((d_5 - d_5).unwrap(), BlockDistance::new(0));
+        assert_eq!((d_5 - d_6).unwrap(), BlockDistance::new(-1));
+
+        assert_eq!((d_5 + d_4).unwrap(), BlockDistance::new(9));
+        assert_eq!((d_5 + d_5).unwrap(), BlockDistance::new(10));
+        assert_eq!((d_5 + d_6).unwrap(), BlockDistance::new(11));
 
-    pub fn checked_add(&self, rhs: u64) -> Option<Self> {
-        self.0.checked_add(rhs).map(BlockHeight::new)
+        assert_eq!(BlockHeight::max() - BlockDistance::new(1), None);
+        assert_eq!(BlockHeight::max() + BlockDistance::new(0), None);
+        assert_eq!(BlockHeight::max() + BlockDistance::new(1), None);
```

Comment:
Would like to see tests similar to these three but with `BlockHeight::new(DistanceIntType::MAX as HeightIntType)` in place of `BlockHeight::max()`